### PR TITLE
[Estuary][PVR][guiinfo] PVR Providers User Interface

### PR DESCRIPTION
**Pull request by ksooo**
We have support for PVR providers in PVR core and PVR client add-ons API for a long time already, thanks to @phunkyfish. This PR brings this core feature to the end user, by implementing respective user interface components and Estuary support.

Screen shots:
 
screenshot00003
screenshot00009
screenshot00011
screenshot00012
screenshot00005
screenshot00008
Screenshot 2024-08-24 at 14 54 55

Changes to the Skinning Engine:
* New window: PVRProviders.xml
* New graphics: DefaultPVRProviders.png, DefaultPVRProvider.png

New GUI info labels: 
* ListItem.MediaProviders, VideoPlayer.MediaProviders, MusicPlayer.MediaProviders - string, containing the names of media providers, separated by commas if multiple. Implemented currently for PVR recordings, channels, EPG (via channels), timers (via channels), but most probably could be implemented for other media as well.

New window IDs:
* "radioproviders"
* "tvproviders"

Runtime-tested on macOS and Android, latest xbmc master, with (pre-release) versions of several PVR Client add-ons, in mixture of add-ons supporting providers and not supporting providers (pvr.demo, pvr.hts, pvr.plutotv)

@phunkyfish can you review (and runtime-test) this?
@jjd-uk  can you please have a look at the Estuary changes?

### CODE REVIEW

File: xbmc/pvr/providers/PVRProvidersPath.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  bool Init(const std::string& path);
+
+  std::string m_path;
+  bool m_isValid{false};
+  bool m_isRoot{false};
+  bool m_isProvider{false};
+  bool m_isChannels{false};
+  bool m_isRecordings{false};
+  Kind m_kind{Kind::UNKNOWN};
+  int m_providerUid{-1};
```
Comment (phunkyfish) [inline xbmc/pvr/providers/PVRProvidersPath.h:57]: Magic number

File: xbmc/pvr/providers/PVRProvidersPath.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  int m_clientId{-1};
```
Comment (phunkyfish) [inline xbmc/pvr/providers/PVRProvidersPath.h:58]: Magic number

File: xbmc/pvr/PVRPathUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  if (filterByClientAndProvider)
+  {
+    clientId = std::atoi(clientIdStr.c_str());
+    providerId = std::atoi(providerIdStr.c_str());
+    return true;
+  }
+  else
+  {
+    clientId = -1;
+    providerId = -1;
```
Comment (phunkyfish) [inline xbmc/pvr/PVRPathUtils.cpp:48]: Magic numbers

File: xbmc/pvr/PVRPathUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    return false;
+  }
+}
+
+std::string GetProviderNameFromPath(const std::string& path)
+{
+  const CURL url{path};
+  int clientId{-1};
+  int providerId{-1};
```
Comment (phunkyfish) [inline xbmc/pvr/PVRPathUtils.cpp:57]: Magic numbers

File: xbmc/pvr/filesystem/PVRGUIDirectory.cpp [hunk 5]
```diff
@@ -309,9 +310,39 @@ bool IsDirectoryMember(const std::string& strDirectory,
...
+
+      if ((m_providerId != PVR_PROVIDER_INVALID_UID) && (item->ClientProviderUid() != m_providerId))
+        return true;
+    }
+    return false;
+  }
+
+private:
+  int m_clientId{-1};
+  int m_providerId{-1};
```
Comment (phunkyfish) [inline xbmc/pvr/filesystem/PVRGUIDirectory.cpp:337]: Magic numbers

**Review by phunkyfish — COMMENTED** [review]
Just some magic numbers after an intial quick review.

**Comment by phunkyfish** [comment]
Looks great, btw. Will do some runtime testing also once the PR builds.

**Comment by ksooo** [comment]
"Problem" with the magic numbers is that I'm hesitant to include the respective addon-devkit header files just to get the define.

**Comment by phunkyfish** [comment]
> "Problem" with the magic numbers is that I'm hesitant to include the respective addon-devkit header files just to get the define.

Maybe just create Kodi core constants to represent them. Sure they could go out of sync but it’s unlikely.

**Comment by ksooo** [comment]
> Sure they could go out of sync 

Really don't like that. But magic number can also go out of sync. Thinking of including the header.

**Comment by ksooo** [comment]
@phunkyfish magic numbers for provider and client ids should be history now. ;-)

**Review by phunkyfish — APPROVED** [review]
Having runtime tested this (great feature BTW!), and reviewed the C++ code, happy to approve

**Comment by ksooo** [comment]
> Let me know when sam's icons are added 

I think I will merge this weekend, no matter I got new icons. Not a show stopper. The current icons are not that bad.

**Comment by ksooo** [comment]
@jjd-uk the new icons are incorporated now. 
image

**Review by jjd-uk — APPROVED** [review]
LGTM

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
jenkins build this please

**Comment by phunkyfish** [comment]
Do you want providers to appear in the left nav with the other PVR windows?

Screenshot 2024-08-30 at 12 58 00

**Comment by ksooo** [comment]
> Do you want providers to appear in the left nav with the other PVR windows?

Honestly, i see no value in this navigation bar which we do not even have for other kodi components. I’d rather remove it instead of extending it. 😉

**Comment by sarbes** [comment]
Please make sure to submit a clean texture next time. The color channels of both textures are broken, which leads to worse image quality, larger on disk size, and larger size in GPU RAM.

**Pull request merged**